# Part 1. Equation of a Slime

How many late days are you using for this assignment?

0

In [20]:
# Imports section
import pandas as pd
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import KFold
from sklearn.preprocessing import PolynomialFeatures
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.neighbors import KNeighborsClassifier
from sklearn.neural_network import MLPClassifier

## 1. Loading the dataset

In [21]:
# Using pandas load the dataset
# Output the first 15 rows of the data
# Display a summary of the table information (number of datapoints, etc.)
dataset = pd.read_csv('science_data_large.csv')
dataset.info()
dataset.head(15)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 3 columns):
 #   Column          Non-Null Count  Dtype  
---  ------          --------------  -----  
 0   Temperature °C  1000 non-null   int64  
 1   Mols KCL        1000 non-null   int64  
 2   Size nm^3       1000 non-null   float64
dtypes: float64(1), int64(2)
memory usage: 23.6 KB


,Temperature °C,Mols KCL,Size nm^3
0,469,647,6.244743e+05
1,403,694,5.779610e+05
2,302,975,6.196847e+05
3,779,916,1.460449e+06
4,901,18,4.325726e+04
5,545,637,7.124634e+05
6,660,519,7.006960e+05
7,143,869,2.718260e+05
8,89,461,8.919803e+04
9,294,776,4.770210e+05


## 2. Splitting the dataset

In [22]:
# Take the pandas dataset and split it into our features (X) and label (y)
X = dataset.drop(columns=['Size nm^3'])
y = dataset['Size nm^3']
# Use sklearn to split the features and labels into a training/test set. (90% train, 10% test)
# For grading consistency use random_state=42 
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.1, train_size=0.9, random_state=42)

## 3. Perform a Linear Regression

In [29]:
# Use sklearn to train a model on the training set
model = LinearRegression()
model.fit(X_train, y_train)
# Create a sample datapoint and predict the output of that sample with the trained model
new_data = pd.DataFrame([[250, 200]], columns=X.columns)
predictions = model.predict(new_data)
print("predictions",predictions)
# Report the score for that model using the default score function property of the SKLearn model, in your own words (markdown, not code) explain what the score means
print("Score:", model.score(X_test, y_test))

# Extract the coefficients and intercept from the model and write an equation for h(x) using actual feature names
coefficients = model.coef_
intercept = model.intercept_

# Construct the equation using the actual variable names
latex_equation = f"h(x) = {intercept:.2f} " + " + ".join(
    [f"+{coeff:.2f} * {feature}" for coeff, feature in zip(coefficients, X.columns)]
).replace("+ +", "+ ")  # Ensuring correct formatting for negative signs

print("Coefficients:", coefficients)
print("Intercept:", intercept)
print("LaTeX Equation:", latex_equation)



predictions [13684.13705673]
Score: 0.8552472077276095
Coefficients: [ 866.14641337 1032.69506649]
Intercept: -409391.47958340764
LaTeX Equation: h(x) = -409391.48 +866.15 * Temperature °C + 1032.70 * Mols KCL


Write the linear equation of a slime: (example equation: $E = mc^2$)

### Hypothesis Function \( h(x) \)

The equation for our trained linear regression model is:

$$
h(x) = \theta_0 + \theta_1 x_1 + \theta_2 x_2
$$

Substituting the learned values:

$$
h(x) = -409391.48 + 866.15 \cdot \text{Temperature °C} + 1032.70 \cdot \text{Mols KCL}
$$



Report on score and explain meaning:

The score of the model, which is 0.8552, represents the coefficient of determination, commonly known as $R^2$ (R-squared). This statistical measure indicates how well the proposed linear regression model explains the variability of the target variable based on the input features.

An $R^2$ value of 0.8552 means that approximately 85.52% of the variance in the target variable can be explained by the features included in the model. This signifies a strong positive correlation between the independent variables and the target variable, implying that the model does a good job of predicting the outcomes based on the provided inputs.

## 4. Use Cross Validation

In [10]:
# Use the cross_val_score function to repeat your experiment across many shuffles of the data
# For grading consistency use n_splits=5 and random_state=42
kf = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores = cross_val_score(model, X, y, cv=kf)
print("Cross-validation scores:", cv_scores)
print("Mean CV score:", cv_scores.mean())
# Report on their finding and their significance



Cross-validation scores: [0.86151889 0.82742341 0.87195173 0.88166206 0.85609101]
Mean CV score: 0.8597294202684646


Write findings here:

The cross-validation implementation provided a series of scores representing the model’s performance across five distinct folds of the dataset.

The Mean Cross-Validation Score calculated from these folds is 0.8597, which indicates the average performance of the model when evaluated on various partitions of the data.

The fact that all the scores are reasonably high (generally above 0.8) indicates that the model is consistently effective at predicting outcomes across different subsets of the dataset. A mean score of 0.8597, close to the previously obtained $R^2$ score of 0.8552, reinforces that the model has strong predictive capabilities.

The relatively small range of scores (from ~0.827 to ~0.882) indicates good stability of the model in its predictions across different data slices. This means the model is less likely to be overfitting or underfitting and thus can be considered robust overall.
Overall, these findings affirm the effectiveness of the regression model while also providing valuable insights into its predictability and reliability in various scenarios.

## 5. Using Polynomial Regression

In [11]:
# Using the PolynomialFeatures library perform another regression on an augmented dataset of degree 2
# Perform k-fold cross validation (as above)
poly = PolynomialFeatures(degree=2)
X_poly = poly.fit_transform(X)
X_t, X_tt, y_t, y_tt = train_test_split(X_poly, y, test_size=0.1, train_size=0.9, random_state=42)
model = LinearRegression()
model.fit(X_poly, y)
kf = KFold(n_splits=5, shuffle=True, random_state=42)
cv_scores_poly = cross_val_score(model, X_poly, y, cv=kf)

print("Polynomial Regression Cross-validation scores:", cv_scores_poly)
print("Mean CV score for Polynomial Regression:", cv_scores_poly.mean())

coefficients_poly = model.coef_
intercept_poly = model.intercept_

feature_names = poly.get_feature_names_out(X.columns)
equation = f"h(x) = {intercept_poly:.2f} " + " ".join(
    [f"+ {coef:.2f} {name}" for coef, name in zip(coefficients_poly, feature_names)]
)

print("Polynomial Regression Equation:")
print(equation)

# Report on the metrics and output the resultant equation as you did in Part 3.



Polynomial Regression Cross-validation scores: [1. 1. 1. 1. 1.]
Mean CV score for Polynomial Regression: 1.0
Polynomial Regression Equation:
h(x) = 0.00 + 0.00 1 + 12.00 Temperature °C + -0.00 Mols KCL + -0.00 Temperature °C^2 + 2.00 Temperature °C Mols KCL + 0.03 Mols KCL^2


Write the polynomial equation of a slime: (example equation: $E = mc^2$)

### Polynomial Regression Equation:

The polynomial regression model can be expressed as:
$$h(x) = 0.00 + 0.00 \cdot 1 + 12.00 \cdot \text{Temperature °C} - 0.00 \cdot \text{Mols KCL} - 0.00 \cdot (\text{Temperature °C})^2$$

$$+ 2.00 \cdot (\text{Temperature °C} \cdot \text{Mols KCL}) + 0.03 \cdot (\text{Mols KCL})^2$$



- The equation indicates that:
  - The term **$12.00 \cdot Temperature °C$** has a strong positive influence on the output variable.
  - The interaction term **$2.00 \cdot (Temperature °C \cdot Mols KCL)$** demonstrates that the effect of Temperature interacts with Mols KCL, suggesting a complex relationship.
  - Several terms have coefficients of **0.00**, indicating that they do not contribute significantly to the model's outcome for these specific data points.

Report on the score and interpret:

### Cross-Validation Scores

**Polynomial Regression Cross-validation scores**:  
- [1.0, 1.0, 1.0, 1.0, 1.0]  

**Mean Cross-Validation Score for Polynomial Regression**:  
- 1.0  

### Interpretation of Findings:

#### Perfect Cross-Validation Scores:
- The fact that all five cross-validation scores are exactly 1.0 indicates that the polynomial regression model perfectly predicted the outcomes for all folds of the dataset.
- This perfect score suggests that the model has an exceptional fit with the data, capturing all the underlying patterns without any errors across the different splits used in cross-validation.

#### Potential Overfitting:
- While achieving a mean score of 1.0 sounds ideal, it could also indicate potential overfitting. The model appears to fit the training data too closely, which may not generalize well to new, unseen data. Additional validation using an independent test set (beyond the k-fold used) would be prudent to assess how the model performs in real-world scenarios.


### Summary
In summary, while the polynomial regression model has achieved outstanding performance metrics with perfect cross-validation scores, one must approach the results with caution regarding overfitting. Future analyses should include testing the model on unseen data to verify its generalizability. This approach will help ensure that the model maintains its performance and is not simply memorizing the training data.


# Part 2. Chronic Kidney Disease Prediction via Classification

Create code and markdown cells as needed to perform classification and report on your results

In [16]:
# Load the dataset. Then train and evaluate the classification models.
dataset_2 = pd.read_csv('ckd_feature_subset.csv')
dataset_2.head(15)
# Assuming 'target' is the column name for the label
X = dataset_2.drop(columns=['Target_ckd'])
y = dataset_2['Target_ckd']
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)


## Results and Conclusion for Classification Experiments

In [17]:

models = {
    'Logistic Regression': LogisticRegression(max_iter=1000),
    'Support Vector Machine': SVC(),
    'k-Nearest Neighbors': KNeighborsClassifier(),
    'Neural Network': MLPClassifier(max_iter=1000)
}
results = {}

for model_name, model in models.items():
    scores = cross_val_score(model, X, y, cv=5)
    results[model_name] = {
        'Mean Accuracy': scores.mean(),
        'Std Dev': scores.std()
    }
results_df = pd.DataFrame(results).T  # Transpose for better visualization
print(results_df)

                        Mean Accuracy   Std Dev
Logistic Regression          0.856129  0.054236
Support Vector Machine       0.928172  0.059993
k-Nearest Neighbors          0.928172  0.031893
Neural Network               0.941290  0.024040


### Performance of Classification Models

The following table summarizes the performance metrics of various classification models evaluated using 5-fold cross-validation:

| Model                     | Mean Accuracy | Std Dev  |
|---------------------------|---------------|----------|
| Logistic Regression        | 0.8561        | 0.0542   |
| Support Vector Machine     | 0.9282        | 0.0600   |
| k-Nearest Neighbors        | 0.9282        | 0.0319   |
| Neural Network             | 0.9413        | 0.0240   |

#### Summary of Results:

- **Logistic Regression**: 
  - **Mean Accuracy**: 0.8561
  - **Standard Deviation**: 0.0542
  - This model shows decent performance but may struggle with complex relationships in the data due to its linear nature.

- **Support Vector Machine**:
  - **Mean Accuracy**: 0.9282
  - **Standard Deviation**: 0.0600
  - Provides a strong performance, effectively classifying data with a non-linear decision boundary.

- **k-Nearest Neighbors**:
  - **Mean Accuracy**: 0.9282
  - **Standard Deviation**: 0.0319
  - Comparable to the SVC model, KNN performs well, though sensitive to the choice of k and the dataset's feature space.

- **Neural Network**:
  - **Mean Accuracy**: 0.9413
  - **Standard Deviation**: 0.0240
  - The best-performing model, indicating that it has a strong capability to learn complex relationships while maintaining stability across folds.

### Conclusion on Findings

**Best Performing Model**:
- The **Neural Network** emerges as the top performer with a mean accuracy of **0.9413**. Its architecture allows it to learn intricate patterns from the dataset effectively.

**Significance of Results**:
- The results demonstrate that while Logistic Regression serves as a solid baseline, the more complex models (SVM, KNN, and Neural Networks) yield substantially better performance. 
- The **Support Vector Machine** and **k-Nearest Neighbors** also perform excellently, both achieving a mean accuracy of **0.9282**. 
- The low standard deviations (especially for the Neural Network) indicate that these models are consistent across different data partitions, suggesting robustness.

However, it is noteworthy to ensure that the Neural Network, despite its high performance, is cross-validated on an independent test set to mitigate overfitting risks. This is crucial for validating that the model generalizes well to unseen data, beyond the training and validation phases used in this assessment.


## Results and Conclusion for Neural Network Experiments

In [18]:
# Define different configurations
nn_configs = {
    'Config 1': MLPClassifier(hidden_layer_sizes=(50,), max_iter=1000),
    'Config 2': MLPClassifier(hidden_layer_sizes=(100,), activation='relu', max_iter=1000),
    'Config 3': MLPClassifier(hidden_layer_sizes=(100, 50), activation='tanh', max_iter=1000),
}

# Evaluate each configuration
nn_results = {}

for config_name, model in nn_configs.items():
    scores = cross_val_score(model, X, y, cv=5)
    nn_results[config_name] = {
        'Mean Accuracy': scores.mean(),
        'Std Dev': scores.std()
    }

# Convert NN results to DataFrame
nn_results_df = pd.DataFrame(nn_results).T
print(nn_results_df)


          Mean Accuracy   Std Dev
Config 1       0.941290  0.024040
Config 2       0.941290  0.024040
Config 3       0.960645  0.024554


### Neural Network Configurations and Results

After evaluating different configurations of the Neural Network model using 5-fold cross-validation, the results are summarized in the following table:

| Configuration | Mean Accuracy | Std Dev  |
|---------------|---------------|----------|
| Config 1      | 0.9413        | 0.0240   |
| Config 2      | 0.9413        | 0.0240   |
| Config 3      | 0.9606        | 0.0246   |

### Summary of Results

1. **Configuration Comparison**:
   - **Config 1**: This configuration consists of a single hidden layer with **50 neurons**. It achieved a mean accuracy of **0.9413** with a standard deviation of **0.0240**.
   - **Config 2**: In this configuration, there is a single hidden layer with **100 neurons** and uses **ReLU** as the activation function. It achieved the same mean accuracy of **0.9413** and a similar standard deviation of **0.0240**.
   - **Config 3**: This configuration includes two hidden layers with **100 neurons** in the first layer and **50 neurons** in the second layer using the **tanh** activation function. It yielded the highest mean accuracy at **0.9606** and a standard deviation of **0.0246**.

2. **Impact of Configuration**:
   - Notably, **Config 3** outperforms both **Config 1** and **Config 2** by a significant margin, highlighting the beneficial impact of deeper architectures in neural networks. The increase from **1 hidden layer** to **2 hidden layers** allows the model to capture more complex relationships within the data, resulting in an improvement in classification performance.
   - The consistent performance (low standard deviation) across all configurations indicates that the models are stable and not overly sensitive to the specific data partitions used in cross-validation.

### Parameters of Influence:
- **Number of Neurons**: While starting with a single layer of **100 neurons** did not differentiate performance significantly from the layer of **50 neurons**, the additional layer in **Config 3** substantially improved results, emphasizing the advantages of deeper networks when appropriately structured.
- **Activation Function**: The activation functions also play a crucial role in determining the effectiveness of each configuration. Although both **Config 2** and **Config 1** had similar performances, exploring the effects of activation functions further could lead to nuanced understanding.

### Conclusion
In summary, Config 3, with its two hidden layers, clearly demonstrated the best performance among the tested configurations, showcasing how deeper architectures can enhance classification tasks in neural networks. The findings suggest that experimenting with depth (number of layers) and number of neurons can yield notable improvements in model accuracy. Future experiments could also explore variations in the learning rate, batch size, and different activation functions to optimize the performance further.
